// @0xcert/scaffold/dist/core/asset-ledger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeneralAssetLedgerAbility;
(function (GeneralAssetLedgerAbility) {
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["CREATE_ASSET"] = 2] = "CREATE_ASSET";
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["REVOKE_ASSET"] = 4] = "REVOKE_ASSET";
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["TOGGLE_TRANSFERS"] = 8] = "TOGGLE_TRANSFERS";
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["UPDATE_ASSET"] = 16] = "UPDATE_ASSET";
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["ALLOW_CREATE_ASSET"] = 32] = "ALLOW_CREATE_ASSET";
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["UPDATE_URI_BASE"] = 64] = "UPDATE_URI_BASE";
    GeneralAssetLedgerAbility[GeneralAssetLedgerAbility["ALLOW_UPDATE_ASSET_IMPRINT"] = 128] = "ALLOW_UPDATE_ASSET_IMPRINT";
})(GeneralAssetLedgerAbility = exports.GeneralAssetLedgerAbility || (exports.GeneralAssetLedgerAbility = {}));
var SuperAssetLedgerAbility;
(function (SuperAssetLedgerAbility) {
    SuperAssetLedgerAbility[SuperAssetLedgerAbility["MANAGE_ABILITIES"] = 1] = "MANAGE_ABILITIES";
})(SuperAssetLedgerAbility = exports.SuperAssetLedgerAbility || (exports.SuperAssetLedgerAbility = {}));
var AssetLedgerCapability;
(function (AssetLedgerCapability) {
    AssetLedgerCapability[AssetLedgerCapability["DESTROY_ASSET"] = 1] = "DESTROY_ASSET";
    AssetLedgerCapability[AssetLedgerCapability["UPDATE_ASSET"] = 2] = "UPDATE_ASSET";
    AssetLedgerCapability[AssetLedgerCapability["REVOKE_ASSET"] = 4] = "REVOKE_ASSET";
    AssetLedgerCapability[AssetLedgerCapability["TOGGLE_TRANSFERS"] = 3] = "TOGGLE_TRANSFERS";
})(AssetLedgerCapability = exports.AssetLedgerCapability || (exports.AssetLedgerCapability = {}));
